Portable RID PublishAOT doesn't work from source-build VMR .NET 9 Preview 5

<!--
Please keep in mind that the GitHub issue tracker is not intended as a general support forum, but for reporting **non-security** bugs and feature requests.

If you believe you have an issue that affects the SECURITY of the platform, please do NOT create an issue and instead email your issue details to secure@microsoft.com. Your report may be eligible for our [bug bounty](https://www.microsoft.com/en-us/msrc/bounty-dot-net-core) but ONLY if it is reported through email.
For other types of questions, consider using [StackOverflow](https://stackoverflow.com).

-->

### Describe the bug

Trying to publish a hello-world console application with a source-build VMR targeting a portable RID doesn't work. Neither does using `--use-current-runtime`.


### To Reproduce
<!--
We ❤ code! Point us to a minimalistic repro project hosted in a GitHub repo, Gist snippet, or other means to see the isolated behavior.

We may close this issue if:
- the repro project you share with us is complex. We can't investigate custom projects, so don't point us to such, please.
- if we will not be able to repro the behavior you're reporting
-->

```
[omajid@828429dd39b9 ~]$ dotnet publish -r linux-x64 /p:PublishAot=true
    /home/omajid/omajid.csproj : error NU1101: Unable to find package runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler. No packages exist with this id in source(s): nuget.org

Restore failed with 1 error(s) in 12.6s
[omajid@828429dd39b9 ~]$ dotnet publish --use-current-runtime /p:PublishAot=true
    /home/omajid/omajid.csproj : error NU1101: Unable to find package runtime.fedora.40-x64.Microsoft.DotNet.ILCompiler. No packages exist with this id in source(s): nuget.org

Restore failed with 1 error(s) in 0.7s
```

### Further technical details
- Include the output of `dotnet --info`

```
[omajid@828429dd39b9 ~]$ dotnet --info
.NET SDK:
 Version:           9.0.100-preview.5.24307.1
 Commit:            35b2c21ea6
 Workload version:  9.0.100-manifests.9b4ca231
 MSBuild version:   17.11.0-preview-24279-02+b963c24ef

Runtime Environment:
 OS Name:     fedora
 OS Version:  40
 OS Platform: Linux
 RID:         fedora.40-x64
 Base Path:   /usr/lib64/dotnet/sdk/9.0.100-preview.5.24307.1/

.NET workloads installed:
Configured to use loose manifests when installing new manifests.
There are no installed workloads to display.

Host:
  Version:      9.0.0-preview.5.24306.7
  Architecture: x64
  Commit:       a5cc707d97

.NET SDKs installed:
  9.0.100-preview.5.24307.1 [/usr/lib64/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 9.0.0-preview.5.24306.11 [/usr/lib64/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 9.0.0-preview.5.24306.7 [/usr/lib64/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_ROOT       [/usr/lib64/dotnet]

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

cc @tmds

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs
@@ -782,19 +782,11 @@ namespace Microsoft.NET.Build.Tasks
 
                 // Get the best RID for the host machine, which will be used to validate that we can run crossgen for the target platform and architecture
                 var runtimeGraph = new RuntimeGraphCache(this).GetRuntimeGraph(RuntimeGraphPath);
-                string hostRuntimeIdentifier = null;
-                // When a non-portable SDK publishes for its non-portable RID specifically, prefer a non-portable host package.
-                string portableSdkRid = !string.IsNullOrEmpty(NETCoreSdkPortableRuntimeIdentifier) ? NETCoreSdkPortableRuntimeIdentifier : RuntimeIdentifier;
-                bool isNonPortablePublish = RuntimeIdentifier == NETCoreSdkRuntimeIdentifier && NETCoreSdkRuntimeIdentifier != portableSdkRid;
-                if (isNonPortablePublish)
-                {
-                    hostRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(runtimeGraph, NETCoreSdkRuntimeIdentifier, packSupportedRuntimeIdentifiers, out bool wasInGraph);
-                }
-                // Otherwise, non-portable SDKs behave the same as portable SDKs.
-                if (hostRuntimeIdentifier == null)
-                {
-                    hostRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(runtimeGraph, portableSdkRid, packSupportedRuntimeIdentifiers, out bool wasInGraph);
-                }
+                // When publishing for the RuntimeIdentifier that matches NETCoreSdkPortableRuntimeIdentifier, prefer NETCoreSdkPortableRuntimeIdentifier for the host.
+                // This makes us use a portable ILCompiler when publishing for a portable RID on a non-portable SDK.
+                string hostRuntimeIdentifier = RuntimeIdentifier == NETCoreSdkPortableRuntimeIdentifier ? NETCoreSdkPortableRuntimeIdentifier : NETCoreSdkRuntimeIdentifier;
+                // Find the best ILCompiler for hostRuntimeIdentifier.
+                hostRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(runtimeGraph, hostRuntimeIdentifier, packSupportedRuntimeIdentifiers, out bool wasInGraph);
                 if (hostRuntimeIdentifier == null)
                 {
                     return ToolPackSupport.UnsupportedForHostRuntimeIdentifier;

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs
@@ -112,6 +112,8 @@ namespace Microsoft.NET.Build.Tasks
         [Required]
         public string NETCoreSdkRuntimeIdentifier { get; set; }
 
+        public string NETCoreSdkPortableRuntimeIdentifier { get; set; }
+
         [Required]
         public string NetCoreRoot { get; set; }
 
@@ -780,7 +782,19 @@ namespace Microsoft.NET.Build.Tasks
 
                 // Get the best RID for the host machine, which will be used to validate that we can run crossgen for the target platform and architecture
                 var runtimeGraph = new RuntimeGraphCache(this).GetRuntimeGraph(RuntimeGraphPath);
-                var hostRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(runtimeGraph, NETCoreSdkRuntimeIdentifier, packSupportedRuntimeIdentifiers, out bool wasInGraph);
+                string hostRuntimeIdentifier = null;
+                // When a non-portable SDK publishes for its non-portable RID specifically, prefer a non-portable host package.
+                string portableSdkRid = !string.IsNullOrEmpty(NETCoreSdkPortableRuntimeIdentifier) ? NETCoreSdkPortableRuntimeIdentifier : RuntimeIdentifier;
+                bool isNonPortablePublish = RuntimeIdentifier == NETCoreSdkRuntimeIdentifier && NETCoreSdkRuntimeIdentifier != portableSdkRid;
+                if (isNonPortablePublish)
+                {
+                    hostRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(runtimeGraph, NETCoreSdkRuntimeIdentifier, packSupportedRuntimeIdentifiers, out bool wasInGraph);
+                }
+                // Otherwise, non-portable SDKs behave the same as portable SDKs.
+                if (hostRuntimeIdentifier == null)
+                {
+                    hostRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(runtimeGraph, portableSdkRid, packSupportedRuntimeIdentifiers, out bool wasInGraph);
+                }
                 if (hostRuntimeIdentifier == null)
                 {
                     return ToolPackSupport.UnsupportedForHostRuntimeIdentifier;

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs
@@ -780,9 +780,14 @@ namespace Microsoft.NET.Build.Tasks
                 var packNamePattern = knownPack.GetMetadata(packName + "PackNamePattern");
                 var packSupportedRuntimeIdentifiers = knownPack.GetMetadata(packName + "RuntimeIdentifiers").Split(';');
 
-                // When publishing for the RuntimeIdentifier that matches NETCoreSdkPortableRuntimeIdentifier, prefer NETCoreSdkPortableRuntimeIdentifier for the host.
-                // This makes us use a portable ILCompiler when publishing for a portable RID on a non-portable SDK.
-                string hostRuntimeIdentifier = RuntimeIdentifier == NETCoreSdkPortableRuntimeIdentifier ? NETCoreSdkPortableRuntimeIdentifier : NETCoreSdkRuntimeIdentifier;
+                // When publishing for the non-portable RID that matches NETCoreSdkRuntimeIdentifier, prefer NETCoreSdkRuntimeIdentifier for the host.
+                // Otherwise prefer the NETCoreSdkPortableRuntimeIdentifier.
+                // This makes non-portable SDKs behave the same as portable SDKs except for the specific case of targetting the non-portable RID.
+                // It also enables the non-portable ILCompiler to be packaged separately from the SDK and
+                // only required when publishing for the non-portable SDK RID.
+                string portableSdkRid = !string.IsNullOrEmpty(NETCoreSdkPortableRuntimeIdentifier) ? NETCoreSdkPortableRuntimeIdentifier : NETCoreSdkRuntimeIdentifier;
+                bool targetsNonPortableSdkRid = RuntimeIdentifier == NETCoreSdkRuntimeIdentifier && NETCoreSdkRuntimeIdentifier != portableSdkRid;
+                string hostRuntimeIdentifier = targetsNonPortableSdkRid ? NETCoreSdkRuntimeIdentifier : portableSdkRid;
                 // Get the best RID for the host machine, which will be used to validate that we can run crossgen for the target platform and architecture
                 var runtimeGraph = new RuntimeGraphCache(this).GetRuntimeGraph(RuntimeGraphPath);
                 hostRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(runtimeGraph, hostRuntimeIdentifier, packSupportedRuntimeIdentifiers, out bool wasInGraph);

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs
@@ -780,12 +780,11 @@ namespace Microsoft.NET.Build.Tasks
                 var packNamePattern = knownPack.GetMetadata(packName + "PackNamePattern");
                 var packSupportedRuntimeIdentifiers = knownPack.GetMetadata(packName + "RuntimeIdentifiers").Split(';');
 
-                // Get the best RID for the host machine, which will be used to validate that we can run crossgen for the target platform and architecture
-                var runtimeGraph = new RuntimeGraphCache(this).GetRuntimeGraph(RuntimeGraphPath);
                 // When publishing for the RuntimeIdentifier that matches NETCoreSdkPortableRuntimeIdentifier, prefer NETCoreSdkPortableRuntimeIdentifier for the host.
                 // This makes us use a portable ILCompiler when publishing for a portable RID on a non-portable SDK.
                 string hostRuntimeIdentifier = RuntimeIdentifier == NETCoreSdkPortableRuntimeIdentifier ? NETCoreSdkPortableRuntimeIdentifier : NETCoreSdkRuntimeIdentifier;
-                // Find the best ILCompiler for hostRuntimeIdentifier.
+                // Get the best RID for the host machine, which will be used to validate that we can run crossgen for the target platform and architecture
+                var runtimeGraph = new RuntimeGraphCache(this).GetRuntimeGraph(RuntimeGraphPath);
                 hostRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(runtimeGraph, hostRuntimeIdentifier, packSupportedRuntimeIdentifiers, out bool wasInGraph);
                 if (hostRuntimeIdentifier == null)
                 {

--- a/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets
@@ -132,6 +132,7 @@ Copyright (c) .NET Foundation. All rights reserved.
                                 KnownWebAssemblySdkPacks="@(KnownWebAssemblySdkPack)"
                                 UsingMicrosoftNETSdkWebAssembly="$(UsingMicrosoftNETSdkWebAssembly)"
                                 NETCoreSdkRuntimeIdentifier="$(NETCoreSdkRuntimeIdentifier)"
+                                NETCoreSdkPortableRuntimeIdentifier="$(NETCoreSdkPortableRuntimeIdentifier)"
                                 NetCoreRoot="$(NetCoreRoot)"
                                 NETCoreSdkVersion="$(NETCoreSdkVersion)">
 

